pennylane/_qubit_device.py

```diff
@@ -478,7 +482,30 @@ def marginal_prob(self, prob, wires=None):
         # The wires provided might not be in consecutive order (i.e., wires might be [2, 0]).
         # If this is the case, we must permute the marginalized probability so that
         # it corresponds to the orders of the wires passed.
-        basis_states = np.array(list(itertools.product([0, 1], repeat=len(device_wires))))
+        num_wires = len(device_wires)
+
+        if 2 < num_wires < 31:
+
+            # The states_to_binary method creates basis states faster (for
+            # larger systems at times over x25 times faster) than the approach
+            # using itertools, at the expense of using more memory.
+            # Due to the large size of the integer arrays for more than 32
+            # bits, memory allocation errors may arise in the states_to_binary
+            # method. Hence we constraint the dtype of the array to 32 bits.
+            # Due to this constraint, an overflow occurs for 31 or more wires,
+            # therefore this approach is used only for fewer wires.
+            # For smaller number of wires speed is comparable to the next
+            # approach, hence we resort to that one for testing purposes
+            states_base_ten = np.arange(2 ** num_wires, dtype=np.int32)
+            basis_states = self.states_to_binary(states_base_ten, num_wires)
+
+        else:
+            # A slower, but less memory intensive method
+            generator_basis_states = itertools.product((0, 1), repeat=num_wires)
+            basis_states = np.fromiter(itertools.chain(*generator_basis_states), dtype=int).reshape(
+                -1, num_wires
+            )
```

Comment:
Is it worth this being its own function, e.g., `generate_basis_states(num_wires)`. We could also document the nice comment above in the docstring.

Reply:
Created it, thanks!